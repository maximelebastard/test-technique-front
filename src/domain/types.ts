export class ApiError extends Error {}

export type Filter = "popular" | "latest" | "watchlist";

export interface Movie {
  id: string;
  title: string;
  poster?: string;
}

export interface MovieDetails {
  id: string;
  genres?: string[];
  directors?: string[];
  rating?: number;
  totalRatings?: number;
  userRating?: number;
  synopsis?: string;
  related?: Movie[];
  inWatchlist?: boolean;
}

export interface TheMovieDbConfig {
  posterBaseUrl: string;
}

export type Paginated<T> = T & {
  page: number;
  indexInPage: number;
};
