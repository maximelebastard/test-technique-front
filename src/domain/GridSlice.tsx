import { createSlice, PayloadAction } from "@reduxjs/toolkit";

/**
 * Slice related to the actions performed on the grid
 */

interface State {
  selectedMovieId?: string;
}

export const initialState: State = {};
export const slice = createSlice({
  name: "grid",
  initialState,
  reducers: {
    showMovieDetails: (state, action: PayloadAction<string | undefined>) => {
      state.selectedMovieId = action.payload;
    }
  }
});

export const { showMovieDetails } = slice.actions;

export default slice.reducer;
