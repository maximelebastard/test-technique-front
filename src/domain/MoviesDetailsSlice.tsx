import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { Movie, MovieDetails } from "./types";

/**
 * Slice related to the movies details data
 */

interface State {
  isConfigFetching: boolean;
  moviesDetails: { [id: string]: MovieDetails };
  fetchingMoviesIds: { [id: string]: boolean };
  fetchError?: string;
}

export const initialState: State = {
  isConfigFetching: false,
  moviesDetails: {},
  fetchingMoviesIds: {}
};

export type PFetchMovieDetailsRequest = {
  id: string;
};
export type PFetchMovieDetailsSuccess = {
  id: string;
  genres: string[];
  directors: string[];
  rating: number;
  totalRatings: number;
  userRating?: number;
  synopsis: string;
  related: Movie[];
  inWatchlist: boolean;
};
export type PFetchMovieDetailsError = PFetchMovieDetailsRequest & {
  error: string;
};

export type PAddWatchlist = {
  id: string;
};
export type PAddWatchlistError = PAddWatchlist & {
  error: string;
};

export type PRateMovie = {
  id: string;
  rating: number;
};
export type PRateMovieError = PRateMovie & {
  error: string;
};

export const slice = createSlice({
  name: "movies_details",
  initialState,
  reducers: {
    fetchMovieDetailsRequest: (
      state,
      action: PayloadAction<PFetchMovieDetailsRequest>
    ) => {
      state.fetchingMoviesIds[action.payload.id] = true;
      state.fetchError = undefined;
    },
    fetchMovieDetailsSuccess: (
      state,
      action: PayloadAction<PFetchMovieDetailsSuccess>
    ) => {
      delete state.fetchingMoviesIds[action.payload.id];
      state.moviesDetails[action.payload.id] = {
        ...state.moviesDetails[action.payload.id],
        directors: action.payload.directors,
        genres: action.payload.genres,
        rating: action.payload.rating,
        totalRatings: action.payload.totalRatings,
        userRating: action.payload.userRating,
        related: action.payload.related,
        synopsis: action.payload.synopsis,
        inWatchlist: action.payload.inWatchlist
      };
      state.fetchError = undefined;
    },
    fetchMovieDetailsError: (
      state,
      action: PayloadAction<PFetchMovieDetailsError>
    ) => {
      delete state.fetchingMoviesIds[action.payload.id];
      state.fetchError = action.payload.error;
    },
    addWatchlistRequest: (state, action: PayloadAction<PAddWatchlist>) => {
      state.moviesDetails[action.payload.id] = {
        ...state.moviesDetails[action.payload.id],
        inWatchlist: true
      };
      state.fetchError = undefined;
    },
    addWatchlistSuccess: state => {
      state.fetchError = undefined;
    },
    addWatchlistError: (state, action: PayloadAction<PAddWatchlistError>) => {
      state.moviesDetails[action.payload.id] = {
        ...state.moviesDetails[action.payload.id],
        inWatchlist: false
      };
      state.fetchError = action.payload.error;
    },
    removeWatchlistRequest: (state, action: PayloadAction<PAddWatchlist>) => {
      state.moviesDetails[action.payload.id] = {
        ...state.moviesDetails[action.payload.id],
        inWatchlist: false
      };
      state.fetchError = undefined;
    },
    removeWatchlistSuccess: state => {
      state.fetchError = undefined;
    },
    removeWatchlistError: (
      state,
      action: PayloadAction<PAddWatchlistError>
    ) => {
      state.moviesDetails[action.payload.id] = {
        ...state.moviesDetails[action.payload.id],
        inWatchlist: true
      };
      state.fetchError = action.payload.error;
    },
    ratingRequest: (state, action: PayloadAction<PRateMovie>) => {
      state.moviesDetails[action.payload.id] = {
        ...state.moviesDetails[action.payload.id],
        userRating: action.payload.rating
      };
      state.fetchError = undefined;
    },
    ratingSuccess: (state, action: PayloadAction<PRateMovie>) => {
      state.moviesDetails[action.payload.id] = {
        ...state.moviesDetails[action.payload.id],
        userRating: action.payload.rating
      };
      state.fetchError = undefined;
    },
    ratingError: (state, action: PayloadAction<PRateMovieError>) => {
      state.moviesDetails[action.payload.id] = {
        ...state.moviesDetails[action.payload.id],
        userRating: undefined
      };
      state.fetchError = action.payload.error;
    }
  }
});

export const {
  fetchMovieDetailsRequest,
  fetchMovieDetailsSuccess,
  fetchMovieDetailsError,
  addWatchlistRequest,
  addWatchlistSuccess,
  addWatchlistError,
  removeWatchlistRequest,
  removeWatchlistSuccess,
  removeWatchlistError,
  ratingRequest,
  ratingSuccess,
  ratingError
} = slice.actions;

export default slice.reducer;
