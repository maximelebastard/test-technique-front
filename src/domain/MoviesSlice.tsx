import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { TheMovieDbConfig, Movie, Paginated, Filter } from "./types";

/**
 * Slice related to the movies data
 */

interface State {
  isConfigFetching: boolean;
  theMovieDbConfig?: TheMovieDbConfig;
  selectedFilter: Filter;
  areMoviesFetching: boolean;
  movies: { [id: string]: Paginated<Movie> };
  currentPage: number;
  totalPages?: number;
  fetchError?: string;
}

export const initialState: State = {
  selectedFilter: "popular",
  currentPage: 1,
  isConfigFetching: false,
  areMoviesFetching: false,
  movies: {}
};

export type PFetchMoviesSuccess = {
  movies: Movie[];
  page: number;
  totalPages: number;
};

export const slice = createSlice({
  name: "movies",
  initialState,
  reducers: {
    initRequest: () => {
      // see how DataSaga handles it
    },
    fetchConfigRequest: state => {
      state.isConfigFetching = true;
      state.fetchError = undefined;
      state.theMovieDbConfig = undefined;
    },
    fetchConfigSuccess: (state, action: PayloadAction<TheMovieDbConfig>) => {
      state.isConfigFetching = false;
      state.theMovieDbConfig = action.payload;
    },
    fetchConfigError: (state, action: PayloadAction<string>) => {
      state.isConfigFetching = false;
      state.fetchError = action.payload;
    },
    fetchMoviesRequest: state => {
      state.areMoviesFetching = true;
      state.fetchError = undefined;
    },
    fetchMoviesSuccess: (state, action: PayloadAction<PFetchMoviesSuccess>) => {
      const { movies: stateMovies, page, totalPages } = action.payload;
      state.areMoviesFetching = false;
      state.totalPages = totalPages;
      const movies = stateMovies.reduce<{ [id: string]: Paginated<Movie> }>(
        (accumulator, movie, currentIndex) => ({
          ...accumulator,
          [movie.id]: { ...movie, page, indexInPage: currentIndex }
        }),
        {}
      );
      state.movies = { ...state.movies, ...movies };
    },
    fetchMoviesError: (state, action: PayloadAction<string>) => {
      state.areMoviesFetching = false;
      state.fetchError = action.payload;
    },
    setFilter: (state, action: PayloadAction<Filter>) => {
      state.selectedFilter = action.payload;
      state.movies = {};
      state.currentPage = 1;
    },
    nextPage: state => {
      state.currentPage += 1;
    }
  }
});

export const {
  initRequest,
  fetchConfigRequest,
  fetchConfigSuccess,
  fetchConfigError,
  fetchMoviesRequest,
  fetchMoviesSuccess,
  fetchMoviesError,
  setFilter,
  nextPage
} = slice.actions;

export default slice.reducer;
