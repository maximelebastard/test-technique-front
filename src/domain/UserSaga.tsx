import { call, put, takeLatest, select } from "redux-saga/effects";
import { PayloadAction } from "@reduxjs/toolkit";
import * as Api from "../api/theMovieDb";
import * as UserSlice from "./UserSlice";
import { RootState } from "store";

function* fetchRequestToken() {
  try {
    const res = yield call(Api.authRequestToken);
    yield put(UserSlice.fetchRequestTokenSuccess(res.request_token));
  } catch (err) {
    yield put(UserSlice.fetchRequestTokenError(err.toString()));
  }
}

function* fetchSessionId() {
  const requestToken = yield select(
    (state: RootState) => state.user.requestToken
  );
  if (!requestToken) return;

  try {
    const res = yield call(Api.authSession, { requestToken });
    if (res.success === true) {
      yield put(UserSlice.fetchSessionIdSuccess(res.session_id));
    } else {
      throw Error("Unsuccessful session Id fetching");
    }
  } catch (err) {
    yield put(UserSlice.fetchSessionIdError(err.toString()));
  }
}

function* fetchAccountId() {
  const sessionId = yield select((state: RootState) => state.user.sessionId);
  if (!sessionId) return;

  try {
    const res = yield call(Api.fetchAccount, { sessionId });
    yield put(UserSlice.fetchAccountIdSuccess(res.id));
  } catch (err) {
    yield put(UserSlice.fetchAccountIdError(err.toString()));
  }
}

function* storeSessionId(action: PayloadAction<string>) {
  window.localStorage.setItem("tmdb_session", action.payload);
}

function* storeAccountId(action: PayloadAction<string>) {
  window.localStorage.setItem("tmdb_account", action.payload);
}

function* UserSaga() {
  yield takeLatest(UserSlice.fetchRequestTokenRequest.type, fetchRequestToken);
  yield takeLatest(UserSlice.fetchSessionIdRequest.type, fetchSessionId);
  yield takeLatest(UserSlice.fetchAccountIdRequest.type, fetchAccountId);
  yield takeLatest(UserSlice.fetchSessionIdSuccess.type, storeSessionId);
  yield takeLatest(UserSlice.fetchAccountIdSuccess.type, storeAccountId);
}

export default UserSaga;
