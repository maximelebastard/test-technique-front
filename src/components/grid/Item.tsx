import React, { useEffect } from "react";
import styles from "./Grid.module.scss";
import Thumbnail from "../thumbnail/Thumbnail.connected";
import Details from "../details/Details";
import { Movie } from "domain/types";

type GridItemProps = {
  movie: Movie;
  showDetails: boolean;
  onMovieSelected: (movieId?: string) => void;
};

const GridItem: React.FC<GridItemProps> = (props: GridItemProps) => {
  const { movie, showDetails, onMovieSelected } = props;

  const detailsRef = React.createRef<HTMLDivElement>();
  const itemRef = React.createRef<HTMLLIElement>();

  function extendDetails() {
    if (itemRef.current) {
      const paddingBottom =
        showDetails && detailsRef.current
          ? `${detailsRef.current.offsetHeight}px`
          : `10px`;
      itemRef.current.style.paddingBottom = paddingBottom;
    }
  }
  useEffect(extendDetails, [showDetails]);

  return (
    <li className={styles.item} ref={itemRef}>
      <Thumbnail
        movie={movie}
        onClick={() => onMovieSelected(showDetails ? undefined : movie.id)}
      />
      {showDetails && (
        <div className={styles.detailsWrapper}>
          <Details
            ref={detailsRef}
            movie={movie}
            onTabChange={() => extendDetails()}
          />
        </div>
      )}
    </li>
  );
};

export default GridItem;
