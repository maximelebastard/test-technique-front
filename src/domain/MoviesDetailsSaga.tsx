import {
  call,
  put,
  all,
  takeLatest,
  takeEvery,
  select
} from "redux-saga/effects";
import { PayloadAction } from "@reduxjs/toolkit";
import * as MoviesDetailsSlice from "./MoviesDetailsSlice";
import * as Api from "../api/theMovieDb";
import { RootState } from "store";
import * as GridSlice from "./GridSlice";

function* fetchMovieDetails(
  action: PayloadAction<MoviesDetailsSlice.PFetchMovieDetailsRequest>
) {
  try {
    const sessionId = yield select((state: RootState) => state.user.sessionId);

    const data = yield all([
      call(Api.fetchMovie, { id: action.payload.id }),
      call(Api.fetchAccountState, { movieId: action.payload.id, sessionId })
    ]);

    const directors = data[0].credits.crew
      .filter((crewMember: any) => crewMember.job === "Director")
      .map((crewMember: any) => crewMember.name);

    yield put(
      MoviesDetailsSlice.fetchMovieDetailsSuccess({
        id: action.payload.id,
        directors: directors.length > 0 ? directors : undefined,
        genres:
          data[0].genres.length > 0
            ? data[0].genres.map((genre: any) => genre.name)
            : undefined,
        synopsis: data[0].overview,
        rating: data[0].vote_average,
        totalRatings: data[0].vote_count,
        userRating: data[1].rated != "false" ? data[1].rated.value : undefined,
        inWatchlist: data[1].watchlist,
        related: data[0].recommendations.results
          .slice(0, 5)
          .map((reco: any) => ({
            id: reco.id,
            title: reco.title,
            poster: reco.poster_path
          }))
      })
    );
  } catch (e) {
    console.error(e);
    yield put(
      MoviesDetailsSlice.fetchMovieDetailsError({
        id: action.payload.id,
        error: e.toString()
      })
    );
  }
}

function* rateMovie(action: PayloadAction<MoviesDetailsSlice.PRateMovie>) {
  try {
    const sessionId = yield select((state: RootState) => state.user.sessionId);

    yield call(Api.rateMovie, {
      sessionId,
      movieId: action.payload.id,
      rating: action.payload.rating
    });

    yield put(
      MoviesDetailsSlice.ratingSuccess({
        id: action.payload.id,
        rating: action.payload.rating
      })
    );
  } catch (e) {
    console.error(e);
    yield put(
      MoviesDetailsSlice.ratingError({
        id: action.payload.id,
        rating: action.payload.rating,
        error: e.toString()
      })
    );
  }
}

function* addToWatchlist(
  action: PayloadAction<MoviesDetailsSlice.PAddWatchlist>
) {
  try {
    const { sessionId, accountId } = yield select(
      (state: RootState) => state.user
    );

    yield call(Api.addToWatchlist, {
      sessionId,
      accountId,
      movieId: action.payload.id,
      inWatchlist: true
    });

    yield put(MoviesDetailsSlice.addWatchlistSuccess());
  } catch (e) {
    console.error(e);
    yield put(
      MoviesDetailsSlice.addWatchlistError({
        id: action.payload.id,
        error: e.toString()
      })
    );
  }
}

function* removeFromWatchlist(
  action: PayloadAction<MoviesDetailsSlice.PAddWatchlist>
) {
  try {
    const { sessionId, accountId } = yield select(
      (state: RootState) => state.user
    );

    yield call(Api.addToWatchlist, {
      sessionId,
      accountId,
      movieId: action.payload.id,
      inWatchlist: false
    });

    yield put(MoviesDetailsSlice.removeWatchlistSuccess());
  } catch (e) {
    console.error(e);
    yield put(
      MoviesDetailsSlice.removeWatchlistError({
        id: action.payload.id,
        error: e.toString()
      })
    );
  }
}

function* handleShowDetails(action: PayloadAction<string | undefined>) {
  if (action.payload) {
    yield put(
      MoviesDetailsSlice.fetchMovieDetailsRequest({ id: action.payload })
    );
  }
}

function* MoviesDetailsSaga() {
  yield takeLatest(
    MoviesDetailsSlice.fetchMovieDetailsRequest.type,
    fetchMovieDetails
  );
  yield takeEvery(GridSlice.showMovieDetails.type, handleShowDetails);
  yield takeEvery(MoviesDetailsSlice.ratingRequest.type, rateMovie);
  yield takeEvery(MoviesDetailsSlice.addWatchlistRequest.type, addToWatchlist);
  yield takeEvery(
    MoviesDetailsSlice.removeWatchlistRequest.type,
    removeFromWatchlist
  );
}

export default MoviesDetailsSaga;
