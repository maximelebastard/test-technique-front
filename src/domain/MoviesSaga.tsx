import { call, put, take, takeLatest, select } from "redux-saga/effects";
import * as MoviesSlice from "./MoviesSlice";
import * as Api from "../api/theMovieDb";
import { RootState } from "store";
import { Movie, Filter } from "./types";

function* fetchConfig() {
  try {
    const data = yield call(Api.fetchConfig);
    yield put(
      MoviesSlice.fetchConfigSuccess({
        posterBaseUrl: `${data.images.secure_base_url}${data.images.poster_sizes[3]}`
      })
    );
  } catch (e) {
    console.error(e);
    yield put(MoviesSlice.fetchConfigError(e));
  }
}

function* fetchMovies() {
  const page = yield select((state: RootState) => state.movies.currentPage);
  const selectedFilter: Filter = yield select(
    (state: RootState) => state.movies.selectedFilter
  );
  const { sessionId, accountId } = yield select(
    (state: RootState) => state.user
  );

  const endpointFromFilter: {
    [filter in Filter]: (p: Api.FetchMoviesParams) => Promise<any>;
  } = {
    popular: Api.fetchPopularMovies,
    latest: Api.fetchLatestMovies,
    watchlist: Api.fetchWatchlistMovies
  };

  try {
    const data = yield call(endpointFromFilter[selectedFilter], {
      page,
      sessionId,
      accountId
    });
    const movies = data.results.map(
      (result: any): Movie => ({
        id: result.id,
        title: result.title,
        poster: result.poster_path
      })
    );
    yield put(
      MoviesSlice.fetchMoviesSuccess({
        page,
        movies,
        totalPages: data.total_pages
      })
    );
  } catch (e) {
    console.error(e);
    yield put(MoviesSlice.fetchMoviesError(e.toString()));
  }
}

function* initializeApp() {
  yield put(MoviesSlice.fetchConfigRequest());
  yield take(MoviesSlice.fetchConfigSuccess.type);
  yield put(MoviesSlice.fetchMoviesRequest());
}

function* handleFilterChange() {
  yield put(MoviesSlice.fetchMoviesRequest());
}

function* fetchNextPage() {
  yield put(MoviesSlice.fetchMoviesRequest());
}

function* MoviesSaga() {
  yield takeLatest(MoviesSlice.initRequest.type, initializeApp);
  yield takeLatest(MoviesSlice.setFilter.type, handleFilterChange);
  yield takeLatest(MoviesSlice.fetchConfigRequest.type, fetchConfig);
  yield takeLatest(MoviesSlice.fetchMoviesRequest.type, fetchMovies);
  yield takeLatest(MoviesSlice.nextPage.type, fetchNextPage);
}

export default MoviesSaga;
