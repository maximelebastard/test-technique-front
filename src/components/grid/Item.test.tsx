import React from "react";
import { shallow } from "enzyme";
import Item from "./Item";
import { Movie } from "domain/types";

const movie: Movie = {
  id: "some_movie",
  title: "Interstellar",
  poster: "/some_poster.jpg"
};

it("renders without crashing", () => {
  shallow(
    <Item
      movie={movie}
      showDetails={true}
      onMovieSelected={movieId => movieId}
    />
  );
});

it("matches the snapshot", () => {
  expect(
    <Item
      movie={movie}
      showDetails={true}
      onMovieSelected={movieId => movieId}
    />
  ).toMatchSnapshot();
});
