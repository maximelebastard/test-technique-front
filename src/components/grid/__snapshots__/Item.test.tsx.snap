// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<GridItem
  movie={
    Object {
      "id": "some_movie",
      "poster": "/some_poster.jpg",
      "title": "Interstellar",
    }
  }
  onMovieSelected={[Function]}
  showDetails={true}
/>
`;
